Update release/deployment RFC (#22)
wellcomecollection/docs: rfcs/013-release_deployment_tracking/v1/README.md

# Version 1 

## Implementation from 24/01/2019

> This solution describes a set of steps to track build, release and deployment but is deliberately agnostic of deployment mechanism. 

Reviewer: nit: "deployment" ~> "deploy"
nit: are you missing a comma between "track" and "build"?

What are we building/releasing/deploying?

> This solution describes a set of steps to track build, release and deployment but is deliberately agnostic of deployment mechanism. 

Reviewer 2: >  but is deliberately agnostic of deployment mechanism. 

Is terraform an example of a build mechanism?

Author: `terraform` is currently our deployment mechanism.

> We make use of SSM to track which release should deploy into which environment. Out terraform stacks read from these SSM parameters to get which images to apply.

Reviewer: nit: "We make use of" ~> "We use"
nit: "should deploy" ~> "should be deployed"
nit: "Out terraform stacks read from these SSM parameters" ~> "Our Terraform code reads from SSM"

Terms defined in this document apply only here.

### Build artifacts

> > **Build artifact:** A container image encapsulating application code that can be configured by environment variable. The artifact _does not_ contain configuration.

Reviewer: Suggestion: move the phrase about environment variable until after the second sentence. It doesn't make sense to talk about how it's configured until we know the image itself doesn't contain configuration.

```suggestion
> **Build artifact:** A Docker image containing our application code. An artifact does not contain configuration; instead, it can be configured by passing environment variables.
```

> When container images intended for release are built they will be added to an ECR repository in AWS, by our build tooling.

Reviewer: Suggestion: use active voice:

```suggestion
Our build tooling creates Docker images and publishes them to an ECR repository in AWS.
```

#### Build artifact SSM Parameters

> The URI of the container image will then be used to update an SSM parameter with a key:

Reviewer: ```suggestion
The URI of the container image in ECR will stored in an SSM parameter with a key:
```

 `/{project_id}/images/{label}/{service_id}`

The attributes are described as follows:

> - **project_id**: an identifier for a project.

Reviewer: Examples?

- **label:** A label for the set of images released, e.g. latest, stable, v2-branch

> 

Reviewer: You haven’t defined `service_id`

 For example:

 `/storage/images/latest/archivist`

> SSM parameters provide a versioned record of build artifacts. SSM allows descriptions to be added to updates, these descriptions should contain the `user_id` of what or who is updating the version

Reviewer: Let’s be clearer about what "versioning" means here – apparently SSM records every version of a parameter, which was news to me.

Do we ever use the description?

Author: > Do we ever use the description?

We do not - I think this is one of those things that was intended but not implemented as per the RFC. So should be removed here.

> This mechanism is provided via a python application, packaged in a docker container distributed via the https://github.com/wellcometrust/dockerfiles repository.

Reviewer: ```suggestion
We update SSM parameters with a Python application, packaged in a Docker container distributed via the https://github.com/wellcomecollection/dockerfiles repository.
```

### Project structure

> > **Project:** A set of services that when composed perform a function.

Reviewer: This definition is incredibly generic. It would help to provide some examples, and maybe this is more specific:

```suggestion
> **Project:** A set of services that are deployed together.
```

> In order to build releases that describe which version of a service to deploy to a particular environment we need a machine readable description of project structure.

Reviewer: This sentence is very long and rambly; you can cut most of it without losing anything.

```suggestion
We store information about the structure of a project in a JSON file, which is named `.wellcome_project` and lives in the root of a repository:
```

The following structure is used:

```json
{

>   "stacks": {

Reviewer: I know this is a real project, but "stack" is such an overloaded term in this context that I'd suggest using another example.
